pkg/services/guardian/accesscontrol_guardian.go

```diff
@@ -107,50 +95,72 @@ func NewAccessControlDashboardGuardianByUID(
 			user:             user,
 			ac:               ac,
 			dashboardService: dashboardService,
+			folderService:    folderService,
 		},
 		dashboard: dashboard,
 	}, nil
 }
 
-// NewAccessControlDashboardGuardianByDashboard creates a dashboard guardian by the provided dashboard.
-// This constructor should be preferred over the other two if the dashboard in available
-// since it avoids querying the database for fetching the dashboard.
-func NewAccessControlDashboardGuardianByDashboard(
-	ctx context.Context, cfg *setting.Cfg, dashboard *dashboards.Dashboard, user identity.Requester,
-	ac accesscontrol.AccessControl, dashboardService dashboards.DashboardService,
+// NewAccessControlDashboardGuardianByFolderUID creates a folder guardian by the provided folderUID.
+func NewAccessControlDashboardGuardianByFolderUID(
+	ctx context.Context, cfg *setting.Cfg, folderUID string, user identity.Requester,
+	ac accesscontrol.AccessControl, dashboardService dashboards.DashboardService, foldersService folder.Service,
 ) (DashboardGuardian, error) {
-	if dashboard != nil && dashboard.IsFolder {
-		return &accessControlFolderGuardian{
-			accessControlBaseGuardian: accessControlBaseGuardian{
-				ctx:              ctx,
-				cfg:              cfg,
-				log:              log.New("folder.permissions"),
-				user:             user,
-				ac:               ac,
-				dashboardService: dashboardService,
-			},
-			folder: dashboards.FromDashboard(dashboard),
-		}, nil
+	var f *folder.Folder
+	if folderUID != "" {
+		q := &folder.GetFolderQuery{
```

Comment:
Why are we fetching the folder given we have the UID? I feel like this should be the caller responsibility to check existence.

Reply:
That's a good point actually. I checked the usage of [folder](https://github.com/grafana/grafana/blob/4e313f0755549b443323f5d4427c5cb17826479e/pkg/services/guardian/accesscontrol_guardian.go#L195) and [dashboard](https://github.com/grafana/grafana/blob/4e313f0755549b443323f5d4427c5cb17826479e/pkg/services/guardian/accesscontrol_guardian.go#L190) for access control guardians, and we seem to only use their UIDs (apart from logging, but we could remove some extra logs). So I think we could even just store UIDs, not dashboards and folders. But that's a bigger change that can be tackled separately.

Reply:
> I feel like this should be the caller responsibility to check existence.

Agreed - but that will be a bigger change. Library elements has been using the dashboard by uid guardian to check for existence. My hope for this PR is to keep functionality the same to ensure this port over is doing okay.